### PR TITLE
Re-implement segnet in MXnet

### PR DESCRIPTION
**Pull request by solin319**
Re-implement segnet in MXnet. 
Details in https://github.com/solin319/incubator-mxnet/tree/solin-patch-segnet/example/segnet
The Camvid test accuracy achieved 83% in segnet_basic network.

Does MXNet need this work?

### CODE REVIEW

**Comment by chinakook** [comment]
Very nice work! At long last, Unpooling is reimplemented in MXNet.

**Comment by piiswrong** [comment]
Does result match original paper?

@zhreshold Could you have a look?

**Comment by zhreshold** [comment]
Nice, can you put some instruction of how to use your operators? @solin319 
And I cannot find the performance in either your repo or the original caffe one.

**Comment by solin319** [comment]
Build MXNet with new pooling and upsampling operators.

# copy new operators to src/operator/
cp segnet/op/* incubator-mxnet/src/operator/
# rebuild MXNet from source
cd incubator-mxnet/
make
cd python/
python setup.py install

File: example/segnet/common/contrib_metrics.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+        for label, pred_label in zip(labels, preds):
+            if pred_label.shape != label.shape:
+                pred_label = mx.ndarray.argmax(pred_label, axis=self.axis)
+            pred_label = pred_label.asnumpy().astype('int32')
+            label = label.asnumpy().astype('int32')
+
+            mx.metric.check_label_shapes(label, pred_label)
+
+            self.sum_metric += (pred_label.flat == label.flat).sum()
```
Comment (zhreshold) [inline example/segnet/common/contrib_metrics.py]: If ignore_label is in pred_label, e.g. ignore_label == 0, this is wrong.

File: example/segnet/common/contrib_metrics.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    ('cross-entropy', 0.57159948348999023)
+    """
+    def __init__(self, eps=1e-12, name='cross-entropy',
+                 output_names=None, label_names=None):
+        super(CrossEntropy, self).__init__(
+            name, eps=eps,
+            output_names=output_names, label_names=label_names)
+        self.eps = eps
+
+    def update(self, labels, preds):
```
Comment (zhreshold) [inline example/segnet/common/contrib_metrics.py]: Why not use the native one in metric.py?

File: example/segnet/data_iter.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        """get two list, each list contains two elements: name and nd.array value"""
+        #img_name= self.f.readline().strip('\n').split("\t")[0].split(' ')
+        img_name = self.names[self.cursor + self.num_parts].split(' ')
+        data = {}
+        label = {}
+        data[self.data_name], label[self.label_name] = self._read_img(img_name[0], img_name[1])
+        return list(data.items()), list(label.items())
+
+    def _read_img(self, img_name, label_name):
+        img = Image.open(os.path.join(self.root_dir, img_name))
```
Comment (zhreshold) [inline example/segnet/data_iter.py]: You can use mx.image.imread instead of PIL package.

File: example/segnet/op/pooling_mask_max_2d.cc [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  std::vector<TShape> out_shape, aux_shape;
+  std::vector<int> out_type, aux_type;
+  CHECK(InferType(in_type, &out_type, &aux_type));
+  CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+  DO_BIND_DISPATCH(CreateOp, param_, (*in_type)[0]);
+}
+
+DMLC_REGISTER_PARAMETER(PoolingMaskParam);
+
+MXNET_REGISTER_OP_PROPERTY(PoolingMask, PoolingMaskProp)
```
Comment (zhreshold) [inline example/segnet/op/pooling_mask_max_2d.cc]: Do you think this operator is reusable to be put into operator/contrib?

File: example/segnet/op/upsampling_mask_max_2d.cc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  std::vector<TShape> out_shape, aux_shape;
+  std::vector<int> out_type, aux_type;
+  CHECK(InferType(in_type, &out_type, &aux_type));
+  CHECK(InferShape(in_shape, &out_shape, &aux_shape));
+  DO_BIND_DISPATCH(CreateOp, param_, (*in_type)[0]);
+}
+
+DMLC_REGISTER_PARAMETER(UpSamplingMaskParam);
+
+MXNET_REGISTER_OP_PROPERTY(UpSamplingMask, UpSamplingMaskProp)
```
Comment (zhreshold) [inline example/segnet/op/upsampling_mask_max_2d.cc]: Do you think this operator is reusable to be put into operator/contrib?

File: example/segnet/score.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+# under the License.
+
+import argparse
+import mxnet as mx
+import time
+import os
+import logging
+import pdb
+import numpy as np
+from PIL import Image
```
Comment (zhreshold) [inline example/segnet/score.py]: Same here, use mx.image

File: example/segnet/segment_one_pic.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+import argparse
+import mxnet as mx
+import time
+import os
+import logging
+import numpy as np
+from PIL import Image
```
Comment (zhreshold) [inline example/segnet/segment_one_pic.py]: Same here.

File: example/segnet/common/contrib_metrics.py (hunk 1 above)
Comment (solin319) [reply example/segnet/common/contrib_metrics.py]:
```
elf.sum_metric -= len(pred_label[pred_label == self.ignore_label])
```
This can solve the problem caused by ignore_label  located in pred_label.

File: example/segnet/common/contrib_metrics.py (hunk 2 above)
Comment (solin319) [reply example/segnet/common/contrib_metrics.py]: 1.The native 'CorssEntropy' in metric.py can only used to pred_label with 2-d.
2. In this example, the pred_label is 4-d and the softmax result located in axis 1. We must swap the softmax result to the last axis.
3. Because of the ignore_label=11, the max number of label is 11, but the size of softmax result is only 0-10. We must add one column data to use the command 
`prob = pred[numpy.arange(label.shape[0]), numpy.int64(label)]`

**Comment by solin319** [comment]
@zhreshold 
Thank you for your suggestions. I have use mx.image to read pic and remove results pngs. But I still have two questions. 
1. I think the operator pooling and up-sampling with mask may used in other pixelwise segmentation networks. Shall I move them to operator/contrib.
2. I use PIL to save numpy array to colorful png. I can't find the interface  in mx.image.

**Comment by zhreshold** [comment]
@solin319 
Yes, move to operator/contrib if they are reusable.
Do you really need to save pics? If so, it's fine to leave it there. 
And BTW, do you think your result can match the paper?

**Comment by solin319** [comment]
@zhreshold 
Yes, the test result can match the paper.
In CamVid data-set, we used 367 for training and 233 for testing like paper.
The train command is `python train_segnet.py --gpus 0,1,2,3 --lr=0.01 --network=segnet_basic`.
The result was shown below.
segnet_basic
We also trained with segnet_basic_with_drop, the result was shown below.
segnet_basic_with_drop
The test accuracy was similar to the paper.

**Comment by zhreshold** [comment]
@solin319 Have you finished? Can you rebase to fix the CI?

**Comment by piiswrong** [comment]
Result on the official site is:

Model | Global Accuracy | Class Accuracy | Mean I/U
-- | -- | -- | --
Segnet-Basic | 82.8% | 62.3% | 46.3%
SegNet (Pretrained Encoder) | 88.6% | 65.9% | 50.2%
SegNet (3.5K dataset) | 86.8% | 81.3% | 69.1%

Which one are you comparing to?

**Comment by solin319** [comment]
The first one segnet-basic, training with 367 pics and test in 233 pics.
@piiswrong

**Comment by piiswrong** [comment]
@zhreshold @winstywang @solin319 Any updates on this?

**Comment by solin319** [comment]
We trained segnet with vgg-bn pretrained parameters and got accuracy 0.89.
segnet_vgg_pretrained

**Comment by piiswrong** [comment]
@zhreshold Could you check if this reproduces the paper?

**Comment by zhreshold** [comment]
screen shot 2018-01-31 at 9 33 54 pm
@solin319 Is this table SOTA as you are results are comparing with?

**Comment by solin319** [comment]
comparing with:

Model | Global Accuracy | Class Accuracy | Mean I/U
-- | -- | -- | --
Segnet-Basic | 82.8% | 62.3% | 46.3%
SegNet (Pretrained Encoder) | 88.6% | 65.9% | 50.2%

http://mi.eng.cam.ac.uk/projects/segnet/tutorial.html

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the JIRA id

Thanks!
